util/image_viewer/slip_viewer/kapton_2019_correction_frame_frame_plugin.py

```diff
@@ -0,0 +1,657 @@
+    try:
+      dc = wx.GCDC(dc)
+    except NotImplementedError:
+      pass
+    dc.SetBrush(wx.Brush(colour, wx.TRANSPARENT))
+    for (lon, lat, place, radius, colour, x_off, y_off, pdata) in data:
+      dc.SetPen(wx.Pen(colour))
+      (x, y) = self._pyslip.ConvertGeo2View((lon, lat))
+      dc.DrawCircle(x, y, 1)
+      if colour == "purple":
+        dc.DrawCircle(x, y, 10)
+      pass
+
+  def _draw_contours_layer(self, dc, data, map_rel):
+    """Draw a layer consisting of differently colored contours.
+
+    dc       the device context to draw on
+    data     a list of dictionaries:
+             {"Paths":[Path, Path, ...], "level":level, "color":color}
+             each Path is an iterable of line segments to be overlaid on the pyslip image
+             level is the contour level of the Paths
+             colour is the colour to be applied to the contour lines at this level
+    map_rel  points relative to map if True, MUST BE TRUE for lightweight
+    """
+
+    assert map_rel is True
+    if len(data) == 0:
+      return
+    scale = 2**self._pyslip.tiles.zoom_level
+
+    # Draw line segments of the appropriate color on the pyslip map object
+    try:
+      dc = wx.GCDC(dc)
+    except NotImplementedError:
+      pass
+    for contour_level in data:
+      path_list = contour_level["Paths"]
+      level = contour_level["level"]
+      colour = wx.Colour(*[(1-decimal)*255 for decimal in contour_level["colour"]])
+      dc.SetBrush(wx.Brush(colour))
+      dc.SetPen(wx.Pen(colour))
+      for path in path_list:
+        segments = path.iter_segments()
+        this_vertex = segments.next()[0]
+        vertex_1 = self._pyslip.ConvertGeo2View(this_vertex) # x, y = Geo2View(lon, lat)
+        while True:
+          try:
+            this_vertex = segments.next()[0]
+            vertex_2 = self._pyslip.ConvertGeo2View(this_vertex) # x, y = Geo2View(lon, lat)
+            dc.DrawLine(vertex_1[0], vertex_1[1], vertex_2[0], vertex_2[1])
+            vertex_1 = vertex_2
+          except StopIteration:
+            break
+
+  def _draw_edge_max_layer(self, dc, data, map_rel):
+    """Draw a layer consisting of two line segments, identifying the Kapton absorption edge and max.
+
+    dc       the device context to draw on
+    data     a list of line segments (lon1, lat1, lon2, lat2)
+    map_rel  points relative to map if True, MUST BE TRUE for lightweight
+    """
+    assert map_rel is True
+    if len(data) == 0:
+      return
+    scale = 2**self._pyslip.tiles.zoom_level
+
+    # Draw line segments on the pyslip map object
+    try:
+      dc = wx.GCDC(dc)
+    except NotImplementedError:
+      pass
+    dc.SetBrush(wx.Brush("black"))
+    dc.SetPen(wx.Pen("black"))
+    for segment in data:
+      lon1, lat1, lon2, lat2 = segment
+      x1, y1 = self._pyslip.ConvertGeo2View((lon1, lat1))
+      x2, y2 = self._pyslip.ConvertGeo2View((lon2, lat2))
+      dc.DrawLine(x1, y1, x2, y2)
+
+  def _map_coords(self, x, y, p):
```

Comment:
This is the new class method that replaces two redundant functions both called "map_coords" inside the DrawMaxEdge and UpdateAbsorptionData.  This new class method is nearly identical to that in the UpdateAbsorptionData function but somewhat different from that in DrawMaxEdge.